crates/sui-framework/sources/priority_queue.move

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Priority queue implemented using a max heap.
+module sui::priority_queue {
+    use std::vector;
+
+    const POP_FROM_EMPTY_HEAP : u64 = 0;
+
+    struct PriorityQueue<T: drop> has store, drop {
+        entries: vector<Entry<T>>,
+    }
+
+    struct Entry<T: drop> has store, drop {
+        priority: u64, // higher value means higher priority and will be popped first
+        value: T,
+    }
+
+    /// Create a new priority queue from the entries.
+    public fun new<T: drop>(entries: vector<Entry<T>>) : PriorityQueue<T> {
+        let len = vector::length(&entries);
+        let i = len / 2;
+        while (i > 0) {
+            i = i - 1;
+            max_heapify_recursive(&mut entries, len, i);
+        };
+        PriorityQueue { entries }
+    }
+
+    /// Pop the entry with the highest priority value.
+    public fun pop_max<T: drop>(pq: &mut PriorityQueue<T>) : (u64, T) {
+        let len = vector::length(&pq.entries);
+        assert!(len > 0, POP_FROM_EMPTY_HEAP);
+        let Entry { priority, value } = vector::remove(&mut pq.entries, 0);
+        max_heapify_recursive(&mut pq.entries, len - 1, 0);
+        (priority, value)
+    }
+
+    /// Insert a new entry into the queue.
+    public fun insert<T: drop>(pq: &mut PriorityQueue<T>, priority: u64, value: T) {
+        vector::push_back(&mut pq.entries, Entry { priority, value});
+        let index = vector::length(&pq.entries) - 1;
+        restore_heap_recursive(&mut pq.entries, index);
+    }
+
+    public fun new_entry<T: drop>(priority: u64, value: T): Entry<T> {
+        Entry { priority, value }
+    }
+
+    public fun create_entries<T: drop>(p: vector<u64>, v: vector<T>): vector<Entry<T>> {
+        let len = vector::length(&p);
+        assert!(vector::length(&v) == len, 0);
+        let res = vector::empty();
+        let i = 0;
+        while (i < len) {
+            let priority = vector::remove(&mut p, 0);
```

Comment:
Can we have duplicate priorities?
If so, are we guaranteed deterministic tie-breaking?

Reply:
Good question. It's going to depend on the position of the items in the original vector, the number of items in the vector, etc. So I would say it is deterministic but hard to predict?